src/c++/vdo/base/vdo.c

```diff
@@ -477,6 +543,379 @@ static int register_vdo(struct vdo *vdo)
+
+	vdo->geometry.unused = 0;
+	vdo->geometry.nonce = nonce;
+	vdo->geometry.bio_offset = 0;
+	vdo->geometry.regions[VDO_INDEX_REGION].id = VDO_INDEX_REGION;
+	vdo->geometry.regions[VDO_INDEX_REGION].start_block = 1;
+	vdo->geometry.regions[VDO_DATA_REGION].id = VDO_DATA_REGION;
+	vdo->geometry.regions[VDO_DATA_REGION].start_block = 1 + config->index_blocks;
+	vdo->geometry.index_config.mem = config->index_memory;
+	vdo->geometry.index_config.sparse = config->index_sparse;
+
+	memcpy((unsigned char *) &(vdo->geometry.uuid), uuid, sizeof(uuid_t));
+
+	return VDO_SUCCESS;
+}
+
+/**
+ * initialize_format_layout() - Initialize the VDO structure so it can be used
+ *                              to write the formatting layout to the disk.
+ *
+ * @vdo: The vdo to initialize.
+ * @error_ptr: The reason for any failure during this call.
+ *
+ * Return: VDO_SUCCESS or an error code.
+ **/
+static int initialize_format_layout(struct vdo *vdo, char **error_ptr)
+{
+	int result;
+	uuid_t uuid;
+	nonce_t nonce = current_time_us();
+
+#ifdef __KERNEL__
+	uuid_gen(&uuid);
+#else
+	uuid_generate(uuid);
+#endif
+	result = initialize_volume_geometry(vdo, nonce, &uuid);
+	if (result != VDO_SUCCESS) {
+		*error_ptr = "Failed to initialize volume geometry";
+		return result;
+	}
+
+
+	result = initialize_component_states(vdo, nonce, error_ptr);
+	if (result != VDO_SUCCESS)
+		return result;
+
+	vdo->states.vdo.state = VDO_NEW;
+	return VDO_SUCCESS;
+}
+
+/**
+ * uninitialize_format_layout() - Unnitialize the format layout
+ *
+ * @vdo: The vdo to initialize.
+ *
+ * Return: VDO_SUCCESS or an error code.
+ **/
+static void uninitialize_format_layout(struct vdo *vdo)
+{
+	memset(&vdo->geometry, 0, sizeof(struct volume_geometry));
+	vdo_destroy_component_states(&vdo->states);
+}
+
+/**
+ * vdo_format() - Format a block device to function as a new VDO. This function
+ *                must be called on a device before a VDO can be loaded for the
+ *                first time. Once a device has been formatted, the VDO can be
+ *                loaded and shut down repeatedly. If a new VDO is desired, this
+ *                function should be called again.
+ *
+ * @vdo        The vdo to format.
+ * @error_ptr: The reason for any failure during this call.
+ *
+ * Return: VDO_SUCCESS or an error
+ **/
+static int __must_check vdo_format(struct vdo *vdo, char **error_ptr)
+{
+	int result;
+
+	result = initialize_format_layout(vdo, error_ptr);
+	if (result != VDO_SUCCESS)
+		return result;
+
+	result = write_format_layout(vdo, error_ptr);
+	uninitialize_format_layout(vdo);
+	return result;
+}
+
+/**
+ * vdo_is_formatted() - Check to see whether the areas vdo will write to during
+ *                      formatting are all zeroes.
+ *
+ * @vdo        The vdo to format.
+ * @formatted: The state of the device. 0 means ok to format, 1 otherwise.
+ *
+ * Return: VDO_SUCCESS or an error
+ **/
+static int __must_check vdo_is_formatted(struct vdo *vdo, int *formatted)
```

Comment:
I think this interface is wrong. read the geometry bokc first (which could report an error), then have a function that checks the data, which can't cause an error. It'll make the error handling much clearer in the caller.